tests/test_graph.py

```diff
@@ -108,6 +110,185 @@ def test_add_dependency_strict_node_dependencies():
+        param_name,
+        e_sig.parameters[param_name].annotation,
+        base.SimplePythonDataFrameGraphAdapter(),
+    )
+
+    assert "d" in nodes
+
+    # adding dependency of c on a should fail because the types are incompatible
+    with pytest.raises(ValueError):
+        graph.add_dependency(
+            nodes["e"],
+            "e",
+            nodes,
+            param_name,
+            f_sig.parameters[param_name].annotation,
+            base.SimplePythonDataFrameGraphAdapter(),
+        )
+
+
+def test_add_dependency_input_nodes_compatible_types():
+    """Tests that if functions request an input that we correctly accept compatible types."""
+    b_sig = inspect.signature(tests.resources.compatible_input_types.b)
+    c_sig = inspect.signature(tests.resources.compatible_input_types.c)
+    d_sig = inspect.signature(tests.resources.compatible_input_types.d)
+
+    nodes = {
+        "b": node.Node.from_fn(tests.resources.compatible_input_types.b),
+        "c": node.Node.from_fn(tests.resources.compatible_input_types.c),
+        "d": node.Node.from_fn(tests.resources.compatible_input_types.d),
+    }
+    nodes["b"]._originating_functions = (tests.resources.compatible_input_types.b,)
+    nodes["c"]._originating_functions = (tests.resources.compatible_input_types.c,)
+    nodes["d"]._originating_functions = (tests.resources.compatible_input_types.d,)
+    # what we want to add
+    param_name = "a"
+
+    # this adds 'a' to nodes
+    graph.add_dependency(
+        nodes["b"],
+        "b",
+        nodes,
+        param_name,
+        b_sig.parameters[param_name].annotation,
+        base.SimplePythonDataFrameGraphAdapter(),
+    )
+
+    assert "a" in nodes
+
+    # this adds 'a' to 'c' as well.
+    graph.add_dependency(
+        nodes["c"],
+        "c",
+        nodes,
+        param_name,
+        c_sig.parameters[param_name].annotation,
+        base.SimplePythonDataFrameGraphAdapter(),
+    )
+
+    # test that we shrink the type to the tighter type
+    assert nodes["a"].type == str
+
+    graph.add_dependency(
+        nodes["d"],
+        "d",
+        nodes,
+        param_name,
+        d_sig.parameters[param_name].annotation,
+        base.SimplePythonDataFrameGraphAdapter(),
+    )
+
+
+def test_add_dependency_input_nodes_compatible_types_order_check():
+    """Tests that if functions request an input that we correctly accept compatible types independent of order."""
+    b_sig = inspect.signature(tests.resources.compatible_input_types.b)
+    c_sig = inspect.signature(tests.resources.compatible_input_types.c)
+    d_sig = inspect.signature(tests.resources.compatible_input_types.d)
+
+    nodes = {
+        "b": node.Node.from_fn(tests.resources.compatible_input_types.b),
+        "c": node.Node.from_fn(tests.resources.compatible_input_types.c),
+        "d": node.Node.from_fn(tests.resources.compatible_input_types.d),
+    }
+    nodes["b"]._originating_functions = (tests.resources.compatible_input_types.b,)
```

Comment:
`.copy_with` should be able to do this more ergonomically

Reply:
it doesn't remove lines of code. So I don't feel like it adds much.